src/Raven.Server/Documents/DocumentDatabase.cs

```diff
@@ -555,40 +648,58 @@ private async Task ExecuteClusterTransactionOneByOne(List<ClusterTransactionComm
 
-                    return;
+                    return false;
                 }
             }
+
+            return true;
         }
 
         private void OnClusterTransactionCompletion(ClusterTransactionCommand.SingleClusterDatabaseCommand command,
-            BatchHandler.ClusterTransactionMergedCommand mergedCommands, Exception exception = null)
+            BatchHandler.ClusterTransactionMergedCommand mergedCommands)
         {
             try
             {
                 var index = command.Index;
                 var options = mergedCommands.Options[index];
-                if (exception == null)
+                Task indexTask = null;
+                if (options.WaitForIndexesTimeout != null)
                 {
-                    Task indexTask = null;
-                    if (options.WaitForIndexesTimeout != null)
-                    {
-                        indexTask = BatchHandler.WaitForIndexesAsync(DocumentsStorage.ContextPool, this, options.WaitForIndexesTimeout.Value,
-                            options.SpecifiedIndexesQueryString, options.WaitForIndexThrow,
-                            mergedCommands.LastChangeVector, mergedCommands.LastTombstoneEtag, mergedCommands.ModifiedCollections);
-                    }
+                    indexTask = BatchHandler.WaitForIndexesAsync(DocumentsStorage.ContextPool, this, options.WaitForIndexesTimeout.Value,
+                        options.SpecifiedIndexesQueryString, options.WaitForIndexThrow,
+                        mergedCommands.LastChangeVector, mergedCommands.LastTombstoneEtag, mergedCommands.ModifiedCollections);
+                }
 
-                    var result = new BatchHandler.ClusterTransactionCompletionResult
-                    {
-                        Array = mergedCommands.Replies[index],
-                        IndexTask = indexTask,
-                    };
-                    ClusterTransactionWaiter.SetResult(options.TaskId, index, result);
-                    _nextClusterCommand = command.PreviousCount + command.Commands.Length;
-                    _lastCompletedClusterTransaction = _nextClusterCommand.Value - 1;
-                    return;
+                var result = new ClusterTransactionCompletionResult
+                {
+                    Array = mergedCommands.Replies[index],
+                    IndexTask = indexTask,
+                };
+                RachisLogIndexNotifications.NotifyListenersAbout(index, null);
+                ServerStore.Cluster.ClusterTransactionWaiter.TrySetResult(options.TaskId, result);
+                _nextClusterCommand = command.PreviousCount + command.Commands.Length;
+                _lastCompletedClusterTransaction = _nextClusterCommand.Value - 1;
+            }
+            catch (Exception e)
+            {
+                // nothing we can do
+                if (_logger.IsInfoEnabled)
```

Comment:
For catastrophic failure, I agree but for `TimeoutException` form `BatchHandler.WaitForIndexesAsync` I think the info is enough.
Or actually, we should handle that totally different in this case...
But this is relevant for V5.2 so may be worth opening a dedicated issue for that.
I opened an issue about `WaitForIndexesAsync` while using cluster-wide transaction so we can handle that there
https://issues.hibernatingrhinos.com/issue/RavenDB-18373
@arekpalinski, what do you think?

Reply:
Catastrophic failures are handled by dedicated `CatastrophicFailureHandler` - so it will be recorded to the log.

Not sure what you mean by "we should handle that totally different in this case...". Can you elaborate?

Reply:
I understood it mistakenly, the catch here doesn't catch the `TimeoutException`.
I don't think there is something that can fail here right now but we can change the log to operation.
I will add it to [RavenDB-18373](https://issues.hibernatingrhinos.com/issue/RavenDB-18373) so it will be fixed for v5.2 as well.

Reply:
ok, great